tests/0083-io_event_cb.c

```diff
@@ -0,0 +1,210 @@
+        TEST_ASSERT(opaque == (void*)0x1234,
+                    "Opaque pointer is not as expected (got: %p)", opaque);
+        mtx_lock(&event_receiver.lock);
+        event_receiver.count += 1;
+        mtx_unlock(&event_receiver.lock);
+}
+
+/**
+ * @brief Wait for one or more events to be received.
+ *        Return 0 if no event was received within the timeout. */
+static int wait_event_cb(int timeout_secs) {
+        int event_count = 0;
+        for (; timeout_secs >= 0; timeout_secs--) {
+                mtx_lock(&event_receiver.lock);
+                event_count = event_receiver.count;
+                event_receiver.count = 0;
+                mtx_unlock(&event_receiver.lock);
+                if (event_count > 0 || timeout_secs == 0)
+                        return event_count;
+                rd_sleep(1);
+        }
+        return 0;
+}
+
+
+int main_0083_io_event_cb (int argc, char **argv) {
+        rd_kafka_conf_t *conf;
+        rd_kafka_topic_conf_t *tconf;
+        rd_kafka_t *rk_p, *rk_c;
+        const char *topic;
+        rd_kafka_topic_t *rkt_p;
+        rd_kafka_queue_t *queue;
+        uint64_t testid;
+        int msgcnt = 100;
+        int recvd = 0;
+        int wait_multiplier = 1;
+        rd_kafka_resp_err_t err;
+        enum {
+                _NOPE,
+                _YEP,
+                _REBALANCE
+        } expecting_io = _REBALANCE;
+
+        testid = test_id_generate();
+        topic = test_mk_topic_name(__FUNCTION__, 1);
+
+        rk_p = test_create_producer();
+        rkt_p = test_create_producer_topic(rk_p, topic, NULL);
+        err = test_auto_create_topic_rkt(rk_p, rkt_p);
+        TEST_ASSERT(!err, "Topic auto creation failed: %s",
+                    rd_kafka_err2str(err));
+
+        test_conf_init(&conf, &tconf, 0);
+        rd_kafka_conf_set_events(conf, RD_KAFKA_EVENT_REBALANCE);
+        test_conf_set(conf, "session.timeout.ms", "6000");
+        test_conf_set(conf, "enable.partition.eof", "false");
+        /* Speed up propagation of new topics */
+        test_conf_set(conf, "metadata.max.age.ms", "5000");
+        test_topic_conf_set(tconf, "auto.offset.reset", "earliest");
+        rk_c = test_create_consumer(topic, NULL, conf, tconf);
+
+        queue = rd_kafka_queue_get_consumer(rk_c);
+
+        test_consumer_subscribe(rk_c, topic);
+
+        rd_kafka_queue_io_event_cb_enable(queue, event_cb, (void *)0x1234);
+
+        /**
+         * 1) Wait for rebalance event
+         * 2) Wait 1 interval (1s) expecting no IO (nothing produced).
+         * 3) Produce half the messages
+         * 4) Expect CB
+         * 5) Consume the available messages
+         * 6) Wait 1 interval expecting no CB.
+         * 7) Produce remaing half
+         * 8) Expect CB
+         * 9) Done.
+         */
+        while (recvd < msgcnt) {
+                TEST_SAY("Waiting for event\n");
+                int event_count = wait_event_cb(1 * wait_multiplier);
```

Comment:
variable definitions must go at the top of the scope, before any function calls.